test/call.c

```diff
@@ -3251,3 +3259,142 @@ int test_call_srtp_tx_rekey(void)
 
 	return err;
 }
+
+
+#ifdef USE_TLS
+int test_call_sni(void)
+{
+	int err = 0;
+	struct fixture fix, *f = &fix;
+	struct dns_server *dns_srv = NULL;
+	struct dnsc *dnsc = NULL;
+	char buri_tls[256], curi_tls[256];
+	const char *dp = test_datapath();
+	char s[256];
+
+	conf_config()->sip.verify_server = true;
+
+	/* Setup Mocking DNS Server */
+	err = dns_server_alloc(&dns_srv, false);
+	err |= dns_server_add_a(dns_srv, "retest.server.org", IP_127_0_0_1);
+	err |= dns_server_add_a(dns_srv, "retest.unknown.org", IP_127_0_0_1);
+	err |= dnsc_alloc(&dnsc, NULL, &dns_srv->addr, 1);
+	err |= net_set_dnsc(baresip_network(), dnsc);
+	TEST_ERR(err);
+
+	fixture_init(f);
+
+	mem_deref(f->a.ua);
+	mem_deref(f->b.ua);
+
+	f->behaviour = BEHAVIOUR_ANSWER;
+
+	re_snprintf(s, sizeof(s), "A <sip:a@retest.client.org;transport=tls>"
+		    ";regint=0;cert=%s/sni/client-interm.pem", dp);
+	err = ua_alloc(&f->a.ua, s);
+	TEST_ERR(err);
+
+	re_snprintf(s, sizeof(s), "B <sip:b@retest.server.org;transport=tls>"
+		    ";regint=0;cert=%s/sni/server-interm.pem", dp);
+	err = ua_alloc(&f->b.ua, s);
+	TEST_ERR(err);
+
+	re_snprintf(s, sizeof(s), "C <sip:c@retest.unknown.org;"
+		    "transport=tls>;regint=0;cert=%s/sni/other-cert.pem", dp);
+	err = ua_alloc(&f->c.ua, s);
+	TEST_ERR(err);
+
+	re_snprintf(buri_tls, sizeof(buri_tls), "sip:b@retest.server.org:%u",
+		    sa_port(&f->laddr_tls));
+	re_snprintf(curi_tls, sizeof(curi_tls), "sip:c@retest.unknown.org:%u",
+		    sa_port(&f->laddr_tls));
+
+	/* 1st test. No CA set. Call from A to B. TLS handshake must fail. */
+	f->b.n_closed = 1;
+
+	err = ua_connect(f->a.ua, 0, NULL, buri_tls, VIDMODE_OFF);
+	TEST_ERR(err);
+
+	err = re_main_timeout(5000);
+	TEST_ERR(err);
+	TEST_ERR(fix.err);
+
+	ASSERT_EQ(0, fix.a.n_incoming);
+	ASSERT_EQ(0, fix.a.n_established);
+	ASSERT_EQ(1, fix.a.n_closed);
+	ASSERT_EQ(0, fix.a.close_scode);
+
+	ASSERT_EQ(0, fix.b.n_incoming);
+	ASSERT_EQ(0, fix.b.n_established);
+	ASSERT_EQ(1, fix.b.n_closed);
+	ASSERT_EQ(0, fix.a.close_scode);
+
+	ASSERT_EQ(0, fix.c.n_incoming);
+	ASSERT_EQ(0, fix.c.n_established);
+	ASSERT_EQ(0, fix.c.n_closed);
+	ASSERT_EQ(0, fix.c.close_scode);
+
+	/* 2nd test. CA set. Call from A to C. TLS handshake must fail because
+	 * certificate of C is selected which is from an unknown CA. */
+	re_snprintf(s, sizeof(s), "%s/sni/root-ca.pem", dp);
+	err = tls_add_cafile_path(uag_tls(), s, NULL);
```

Comment:
We may also set `sip_certificate` to a wrong certificate in order to check if the account certificate is chosen by SNI.

Reply:
Good idea. Done.
The global `sip_certificate` is set at the beginning of the test. The final call must still succeed because SNI chooses the correct UA certificate and does not choose the global certificate.